### PR TITLE
Fix performance problem in MutableList.subList() and implement similar optimization as ArrayList.subList()

### PR DESCRIPTION
**Pull request by motlin**
I couldn't figure out how to edit #1620 so I stacked this on top. Please take a look @yoa1226

My changes are in the second commit: a118fd26e5d0055da43e1d9228679f6b3e771e8b

### CODE REVIEW

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/list/mutable/MutableListTestCase.java [hunk 1]
```diff
@@ -95,24 +95,4 @@ default void MutableList_sortThis_comparator()
...
-
-        sublist2.add("X");
-
-        assertEquals(Lists.immutable.with("A", "B", "X", "C"), sublist);
-        assertEquals(Lists.immutable.with("A", "B", "X"), sublist2);
-
-        assertEquals("B", sublist2.remove(1));
-        assertEquals(Lists.immutable.with("A", "X", "C"), sublist);
-        assertEquals(Lists.immutable.with("A", "X"), sublist2);
-    }
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/list/mutable/MutableListTestCase.java:117]: I moved this old test up to ListTestCase to go with the new ones.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/list/mutable/UnmodifiableMutableListTestCase.java [hunk 2]
```diff
@@ -50,11 +50,4 @@ default void MutableList_sortThis_comparator()
         assertThrows(UnsupportedOperationException.class, () ->
                 this.newWith(5, 1, 4, 2, 3).sortThis(Comparators.reverseNaturalOrder()));
     }
-
-    @Override
-    @Test
-    default void MutableList_subList_subList_remove()
-    {
-        assertThrows(UnsupportedOperationException.class, () -> this.newWith().subList(0, 0).remove(new Object()));
-    }
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/list/mutable/UnmodifiableMutableListTestCase.java:59]: This is now covered in FixedSizeListTestCase

**Comment by yoa1226** [comment]
> I couldn't figure out how to edit #1620 so I stacked this on top. Please take a look @yoa1226
> 
> My changes are in the second commit: [a118fd2](https://github.com/eclipse/eclipse-collections/commit/a118fd26e5d0055da43e1d9228679f6b3e771e8b)

@motlin I learn a lot from your code.  I have two question about above.

1. how to merge code? I mean if I need to merge two commits to one. Which pull request do we need ?
2. Do we need resolve performance problem about `clear` method in `SubList` class In this issue?

**Comment by yoa1226** [comment]
> If you like the changes you can fetch this commit and add it to your own branch. No need to squash them. Alternately if we're both happy with this I can just merge this PR.

Yeah, I think we can merge this PR.



> No, it's just an idea for later that we could track with a new issue.

I can continue to do this.

**Comment by yoa1226** [comment]
I'm wondering if we can make the commit message of second commit clear.

**Comment by motlin** [comment]
> I'm wondering if we can make the commit message of second commit clear.

I improved the second commit message. I'm letting the builds rerun, and then I'll plan on merging this PR.

**Pull request merged**